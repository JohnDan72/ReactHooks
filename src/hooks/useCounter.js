import { useState } from "react"


export const useCounter = ( initialState = 3 ) => {
    // if(initialState <= 0 || initialState > 20){
    //     initialState = 0;
    // }

    const [state, setState] = useState(initialState);
    const increment = (value = 1) => setState(state + value);
    const decrement = (value = 1) => setState(state - value);
    
    // const increment = ( value = 1 ) => {
    //     if(state + value <= 20) {setState(state + value)}
    // };
    // const decrement = ( value = 1 ) => {
    //     if(state - value >= 1) {setState(state - value)}
    // };
    const reset = () => setState(1);

    return {
        state,
        increment,
        decrement,
        reset,
    }
}