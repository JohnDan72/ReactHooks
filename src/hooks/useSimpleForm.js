import { useState } from 'react';



export const useSimpleForm = (initialState = {}) => {

    const [formValue, setFormValue] = useState(initialState);

    const handleInputChange = ({ name , value }) => {
        setFormValue({
            ...formValue,
            [name]: value
        });
    }

    return {
        formValue,
        setFormValue,
        handleInputChange,
    };
}


