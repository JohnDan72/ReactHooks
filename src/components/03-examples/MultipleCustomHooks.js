import React from 'react';
import { Button, ButtonToolbar, Col, Divider, FlexboxGrid, Message } from 'rsuite';
import { useFetch } from '../../hooks/useFetch';
// import PropTypes from 'prop-types';

// icons
import { Icon } from '@rsuite/icons';
import { RiReactjsLine } from "react-icons/ri";
import { useCounter } from '../../hooks/useCounter';


const propTypes = {};
const defaultProps = {};


const MultipleCustomHooks = () => {
    const { increment, decrement, state: counter } = useCounter(3);
    const { data, loading } = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`);

    const { author, quote } = !!data && data[0]; //doble negaci√≥n para return 'false' en lugar de 'null'
    console.log(data);


    return (
        <>
            <h3>Use Fetch (Breaking bad API)</h3>

            <Divider></Divider>
            {/* <Icon as={RiReactjsLine} spin size="4em" /> */}

            <FlexboxGrid className="my-4" justify="center" align="middle">
                {
                    (loading)
                        ?
                        (
                            <FlexboxGrid.Item as={Col} >
                                <Icon as={RiReactjsLine} spin size="4em" />
                            </FlexboxGrid.Item>

                        )
                        :
                        (
                            <>
                                <FlexboxGrid.Item as={Col} xs={12}>
                                    <Message showIcon type="success" header="Success">
                                        <h5>Author: {author}</h5>
                                        <h6>Quote: {quote}</h6>
                                    </Message>
                                </FlexboxGrid.Item>
                                <FlexboxGrid.Item as={Col} >
                                    <ButtonToolbar>
                                        <Button color="yellow"
                                            appearance="ghost"
                                            onClick={(e) => decrement()}
                                        >
                                            Anterior
                                        </Button>
                                        <Button color="blue"
                                            appearance="ghost"
                                            onClick={(e) => increment()}
                                        >
                                            Siguiente
                                        </Button>

                                    </ButtonToolbar>
                                </FlexboxGrid.Item>

                            </>
                        )
                }
            </FlexboxGrid>
        </>
    );
}

MultipleCustomHooks.propTypes = propTypes;
MultipleCustomHooks.defaultProps = defaultProps;

export default MultipleCustomHooks;