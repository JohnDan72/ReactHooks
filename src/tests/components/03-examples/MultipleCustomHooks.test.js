import { shallow } from "enzyme"
import MultipleCustomHooks from "../../../components/03-examples/MultipleCustomHooks"
import { useFetch } from "../../../hooks/useFetch";
import { useCounter } from "../../../hooks/useCounter";

jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');


describe('Pruebas MultipleCustomHooks', () => {

    beforeEach(() => {
        useCounter.mockReturnValue({
            state: 1,
            increment: () => { },
            decrement: () => { },
            reset: () => { },
        });
    });

    test('debe renderizar el component ', () => {

        useFetch.mockReturnValue({
            data: null,
            loading: true
        });
        const wrapper = shallow(<MultipleCustomHooks />);
        expect(wrapper).toMatchSnapshot();
    });

    test('debe mostrar la info ', () => {

        const [author, quote] = ['Juan', 'Hola mundo']
        useFetch.mockReturnValue({
            data: [
                {
                    author,
                    quote
                }
            ],
            loading: false,
            error: null
        });
        const wrapper = shallow(<MultipleCustomHooks />);

        const iconTags = wrapper.find('Icon');
        const h5Tag = wrapper.find('h5');
        const h6Tag = wrapper.find('h6');

        expect(wrapper).toMatchSnapshot();
        expect(iconTags.exists()).toBe(false);
        expect(h5Tag.text().includes(author)).toBe(true);
        expect(h6Tag.text().includes(quote)).toBe(true);
    });


})