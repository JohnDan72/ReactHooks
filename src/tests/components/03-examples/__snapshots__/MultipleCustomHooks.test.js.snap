// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas MultipleCustomHooks debe mostrar la info  1`] = `
<Fragment>
  <h3>
    Use Fetch (Breaking bad API)
  </h3>
  <Divider />
  <FlexboxGrid
    align="middle"
    className="my-4"
    justify="center"
  >
    <FlexboxGridItem
      as={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "propTypes": Object {
            "as": [Function],
            "className": [Function],
            "classPrefix": [Function],
            "lg": [Function],
            "lgHidden": [Function],
            "lgOffset": [Function],
            "lgPull": [Function],
            "lgPush": [Function],
            "md": [Function],
            "mdHidden": [Function],
            "mdOffset": [Function],
            "mdPull": [Function],
            "mdPush": [Function],
            "sm": [Function],
            "smHidden": [Function],
            "smOffset": [Function],
            "smPull": [Function],
            "smPush": [Function],
            "xs": [Function],
            "xsHidden": [Function],
            "xsOffset": [Function],
            "xsPull": [Function],
            "xsPush": [Function],
          },
          "render": [Function],
        }
      }
      xs={12}
    >
      <Message
        header="Success"
        showIcon={true}
        type="success"
      >
        <h5>
          Author: 
          Juan
        </h5>
        <h6>
          Quote: 
          Hola mundo
        </h6>
      </Message>
    </FlexboxGridItem>
    <FlexboxGridItem
      as={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "propTypes": Object {
            "as": [Function],
            "className": [Function],
            "classPrefix": [Function],
            "lg": [Function],
            "lgHidden": [Function],
            "lgOffset": [Function],
            "lgPull": [Function],
            "lgPush": [Function],
            "md": [Function],
            "mdHidden": [Function],
            "mdOffset": [Function],
            "mdPull": [Function],
            "mdPush": [Function],
            "sm": [Function],
            "smHidden": [Function],
            "smOffset": [Function],
            "smPull": [Function],
            "smPush": [Function],
            "xs": [Function],
            "xsHidden": [Function],
            "xsOffset": [Function],
            "xsPull": [Function],
            "xsPush": [Function],
          },
          "render": [Function],
        }
      }
    >
      <ButtonToolbar>
        <Button
          appearance="ghost"
          color="yellow"
          onClick={[Function]}
        >
          Anterior
        </Button>
        <Button
          appearance="ghost"
          color="blue"
          onClick={[Function]}
        >
          Siguiente
        </Button>
      </ButtonToolbar>
    </FlexboxGridItem>
  </FlexboxGrid>
</Fragment>
`;

exports[`Pruebas MultipleCustomHooks debe renderizar el component  1`] = `
<Fragment>
  <h3>
    Use Fetch (Breaking bad API)
  </h3>
  <Divider />
  <FlexboxGrid
    align="middle"
    className="my-4"
    justify="center"
  >
    <FlexboxGridItem
      as={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "propTypes": Object {
            "as": [Function],
            "className": [Function],
            "classPrefix": [Function],
            "lg": [Function],
            "lgHidden": [Function],
            "lgOffset": [Function],
            "lgPull": [Function],
            "lgPush": [Function],
            "md": [Function],
            "mdHidden": [Function],
            "mdOffset": [Function],
            "mdPull": [Function],
            "mdPush": [Function],
            "sm": [Function],
            "smHidden": [Function],
            "smOffset": [Function],
            "smPull": [Function],
            "smPush": [Function],
            "xs": [Function],
            "xsHidden": [Function],
            "xsOffset": [Function],
            "xsPull": [Function],
            "xsPush": [Function],
          },
          "render": [Function],
        }
      }
    >
      <Icon
        as={[Function]}
        fill="currentColor"
        height="1em"
        size="4em"
        spin={true}
        width="1em"
      />
    </FlexboxGridItem>
  </FlexboxGrid>
</Fragment>
`;
